---
title: "Query API"
output: html_document
---

**How do I use the query field?**

The [mygene.info](http://mygene.info/) API is used to make the data more searchable/filterable. 
The examples below demonstrate queries.


| Example query text      | Explanation
|-------------------------|--------------------------------------------------------------------------------
|`ids=1017,1018` or `ids=695,ENSG00000123374` | Use list of NCBI or ENSEMBL gene ids to select genes and corresponding proteins
|`tumor suppressor`      | Use keyword or concept to select relevant genes/proteins
|`name:insulin`          | Use gene name to select a gene/protein
|`GO:0030070`            | Use Gene Ontology term [insulin processing](http://amigo.geneontology.org/amigo/term/GO:0030070) to select genes/proteins
|`refseq:NM_001798`      | Use RNA sequence to select corresponding gene/protein
|`refseq:NP_439892`      | Use protein sequence to select corresponding gene/protein
|`chrX:151,073,054-151,383,976` | Use genomic range to select genes/proteins (use with or without commas). Note that `chrX` alone is not a valid query.  

<br/>
You can use AND/OR/NOT boolean operators and grouping to form complicated queries:

Example query text                       | Operator/grouping
-----------------------------------------|---------------------------------------------------------------
`tumor AND suppressor`                   | AND operator
`CDK2 OR BTK`                            | OR operator
`"tumor suppressor" NOT receptor`        | NOT operator
`(interleukin OR insulin) AND receptor`  | grouping with parentheses

